### PR TITLE
create advisories for stream and consumer add/delete/modify

### PR DESCRIPTION
**Pull request by ripienaar**
We now publish advisories when streams and consumers are added,
deleted and modified

Also rework how TypedEvents are created to be easier to use

Signed-off-by: R.I.Pienaar <rip@devco.net>

/cc @nats-io/core

### CODE REVIEW

**Comment by ripienaar** [comment]
This relates to https://github.com/nats-io/jetstream/issues/201

File: server/events.go [hunk 1]
```diff
@@ -222,6 +216,59 @@ type serverUpdate struct {
 	ltime time.Time
 }
 
+const unknownTypedMessageType = "io.nats.unknown_message"
+
+func setTypedEvent(e TypedEvent, idgen ...func() string) TypedEvent {
```
Comment (ripienaar) [inline server/events.go:221]: while this is a pain to maintain, its a lot less painful than having to remember which constant to use and fill in the various TypedEvent fields every time one wants to send an advisory.

If you like this pattern I can do the same with the API stuff where we have all these

```go
var resp = JSApiAccountInfoResponse{ApiResponse: ApiResponse{Type: JSApiAccountInfoResponseType}}
```

that you did not like.

Comment (derekcollison) [reply server/events.go:221]: Might be a better approach.

File: server/consumer.go [hunk 2]
```diff
@@ -504,9 +474,54 @@ func (mset *Stream) AddConsumer(config *ConsumerConfig) (*Consumer, error) {
...
+
+	// can be nil during shutdown, locks are held in the caller
+	if o.mset != nil && o.mset.sendq != nil {
+		subj := JSAdvisoryConsumerDeletedT + "." + o.stream + "." + o.name
+		o.mset.sendq <- &jsPubMsg{subj, subj, _EMPTY_, j, nil, 0}
+	}
+}
+
+func (o *Consumer) sendCreateAdvisory() {
+	o.mu.Lock()
```
Comment (derekcollison) [inline server/consumer.go:503]: Not critical path, can pair with defer unlock.

File: server/consumer.go [hunk 3]
```diff
@@ -504,9 +474,54 @@ func (mset *Stream) AddConsumer(config *ConsumerConfig) (*Consumer, error) {
...
+
+	e := setTypedEvent(&JSConsumerActionAdvisory{
+		Stream:   o.stream,
+		Consumer: o.name,
+		Action:   CreateEvent,
+	})
+
+	j, err := json.MarshalIndent(e, "", "  ")
+	if err != nil {
+		o.mu.Unlock()
```
Comment (derekcollison) [inline server/consumer.go:513]: Then can remove this.

File: server/consumer.go [hunk 4]
```diff
@@ -951,17 +961,12 @@ func (o *Consumer) incDeliveryCount(sseq uint64) uint64 {
 }
 
 func (o *Consumer) notifyDeliveryExceeded(sseq, dcount uint64) {
-	e := &ConsumerDeliveryExceededAdvisory{
-		TypedEvent: TypedEvent{
-			Type: ConsumerDeliveryExceededAdvisoryType,
-			ID:   nuid.Next(),
-			Time: time.Now().UTC(),
-		},
+	e := setTypedEvent(&JSConsumerDeliveryExceededAdvisory{
```
Comment (derekcollison) [inline server/consumer.go:964]: Weird now that I am seeing it I prefer the old style.

File: server/events.go (hunk 1 above)
Comment (ripienaar) [reply server/events.go:221]: OK, will do a follow up PR for those.

File: server/consumer.go [hunk 5]
```diff
@@ -1503,21 +1508,30 @@ func stopAndClearTimer(tp **time.Timer) {
...
+func (o *Consumer) stop(dflag, doSignal, advisory bool) error {
 	o.mu.Lock()
 	mset := o.mset
 	if mset == nil {
 		o.mu.Unlock()
 		return nil
 	}
+
+	if advisory {
+		o.sendDeleteAdvisoryLocked()
```
Comment (derekcollison) [inline server/consumer.go:1532]: stop here does not always mean delete. IIRC when you shutdown a server but expect to restore the consumer state on restart this is called with dflag false. So maybe just need

```go
if dflag && advisory {}
```

File: server/events.go (hunk 1 above)
Comment (derekcollison) [reply server/events.go:221]: I seem to prefer after looking at actual code the old version.

File: server/jetstream_api.go [hunk 6]
```diff
@@ -146,6 +144,21 @@ const (
 	// JetStreamAdvisoryConsumerMaxDeliveryExceedPre is a notification published when a message exceeds its delivery threshold.
 	JSAdvisoryConsumerMaxDeliveryExceedPre = "$JS.EVENT.ADVISORY.CONSUMER.MAX_DELIVERIES"
 
+	// JSAdvisoryStreamCreatedT notification that a stream was created
+	JSAdvisoryStreamCreatedT = "$JS.EVENT.ADVISORY.STREAM.CREATED"
```
Comment (derekcollison) [inline server/jetstream_api.go:148]: T suffix usually preserved for templates that have %s or %d's in them.

Comment (derekcollison) [reply server/jetstream_api.go:148]: Should these have stream and optionally consumer in the subject for authorization?

File: server/stream.go [hunk 7]
```diff
@@ -276,6 +343,7 @@ func (mset *Stream) Delete() error {
 
 // Update will allow certain configuration properties of an existing stream to be updated.
 func (mset *Stream) Update(config *StreamConfig) error {
+	fmt.Printf("!!! updating\n")
```
Comment (derekcollison) [inline server/stream.go:346]: left over debug..

File: server/events.go (hunk 1 above)
Comment (ripienaar) [reply server/events.go:221]: OK :) will remove

File: server/jetstream_api.go (hunk 6 above)
Comment (ripienaar) [reply server/jetstream_api.go:148]: I use them like this `subj := JSAdvisoryConsumerDeletedT + "." + o.stream + "." + o.name`, so not printf style but still not the full end result, can make it sprintf if you prefer

Comment (derekcollison) [reply server/jetstream_api.go:148]: Let's do that for consistency, also more efficient probably garbage wise..

File: server/jetstream_api.go [hunk 8]
```diff
@@ -146,6 +146,21 @@ const (
...
+	JSAdvisoryStreamDeletedPre = "$JS.EVENT.ADVISORY.STREAM.DELETED"
+
+	// JSAdvisoryStreamUpdatedPre notification that a stream was updated
+	JSAdvisoryStreamUpdatedPre = "$JS.EVENT.ADVISORY.STREAM.UPDATED"
+
+	// JSAdvisoryConsumerCreatedPre notification that a template created
+	JSAdvisoryConsumerCreatedPre = "$JS.EVENT.ADVISORY.CONSUMER.CREATED"
+
+	// JSAdvisoryConsumerDeletedPre notification that a template deleted
+	JSAdvisoryConsumerDeletedPre = "$JS.EVENT.ADVISORY.CONSUMER.DELETED"
```
Comment (ripienaar) [inline server/jetstream_api.go:162]: I went with the `xPre` pattern we used here for other advisories where we're also just doing string concats

File: server/consumer.go [hunk 9]
```diff
@@ -504,9 +476,62 @@ func (mset *Stream) AddConsumer(config *ConsumerConfig) (*Consumer, error) {
...
+			Time: time.Now().UTC(),
+		},
+		Stream:   o.stream,
+		Consumer: o.name,
+		Action:   DeleteEvent,
+	}
+
+	j, err := json.MarshalIndent(e, "", "  ")
+	if err != nil {
+		o.mu.Unlock()
```
Comment (derekcollison) [inline server/consumer.go:498]: Dangling unlock.

File: server/consumer.go [hunk 10]
```diff
@@ -1503,21 +1528,30 @@ func stopAndClearTimer(tp **time.Timer) {
...
-func (o *Consumer) stop(dflag, doSignal bool) error {
+func (o *Consumer) stop(dflag, doSignal, advisory bool) error {
 	o.mu.Lock()
 	mset := o.mset
 	if mset == nil {
 		o.mu.Unlock()
 		return nil
 	}
+
+	if dflag && advisory {
```
Comment (derekcollison) [inline server/consumer.go:1551]: Looking at code, do we need advisory flag or just send if dflag true?

File: server/consumer.go (hunk 9 above)
Comment (ripienaar) [reply server/consumer.go:498]: doh, thanks

File: server/consumer.go (hunk 10 above)
Comment (ripienaar) [reply server/consumer.go:1551]: yeah, see the calls `deleteWithoutAdvisory` where we delete a consumer if right after adding it we discover it should have been gone, so there you dont want to notify deleting one that effectively didnt even finish creating

**Pull request merged**